In [123]:
import pandas as pd

In [124]:
raw_data = pd.read_csv("path", parse_dates=["Purchased At"])
print(raw_data.columns)
print(raw_data.shape)
raw_data.head()

Index(['Access Code', 'Name', 'Gender', 'Age', 'Height', 'Initial Weight',
       'Lowest Weight', 'Target Weight', 'Product Name', 'Status', 'Price',
       'Purchased At', 'Payment Type', 'Channel'],
      dtype='object')
(10000, 14)


Access Code Name  Gender   Age  Height  Initial Weight  Lowest Weight  \
0    Y9RY2VSI  김승혜  FEMALE  25.0   172.0            66.9           65.8   
1    3GTN3S3B  허승준    MALE  26.0   176.0            70.0            NaN   
2    6B0IG276  이지민  FEMALE  23.0   171.0            98.0            NaN   
3    EMGRU2MO  장설윤  FEMALE  20.0   160.0            70.7            NaN   
4    1ELG96TX  서성빈  FEMALE  28.0   165.0            55.5            NaN   

   Target Weight         Product Name     Status   Price  \
0      55.000000          눔 체중감량 프로그램  completed  112500   
1      65.000000          눔 체중감량 프로그램  completed   44780   
2      91.140000  눔 체중감량 프로그램 (천원 체험)  completed  132000   
3      53.000000  눔 체중감량 프로그램 (천원 체험)  completed  112500   
4      51.615002          눔 체중감량 프로그램  completed   44780   

             Purchased At Payment Type   Channel  
0 2017-04-14 19:03:29.976    Recurring    others  
1 2017-05-23 20:53:54.368    Recurring    others  
2 2017-08-23 23:39:21.840    Recurring  facebook  
3 2017-08-28 20:18:22.824    Recurring     naver  
4 2017-05-07 17:50:30.944    Recurring  facebook

## 데이터 정리하기


In [125]:
columns = ['Access Code', 
    'Name', 'Gender', 'Age', 'Height', 'Initial Weight', 
    'Lowest Weight', 'Target Weight', 'Status', 'Price', 
    'Purchased At', 'Channel']

data = raw_data[columns].copy()
#copy를 쓰면 data를 고쳐도 raw_data에는 영향이 안감. 그러나 대용량 데이터를 쓸때는 용량이 커지기때문ㅇ
#카피안쓰고 걍 데이터 써야할 것임.

data = data.set_index("Access Code")

print(data.shape)
data.head()


(10000, 11)


Name  Gender   Age  Height  Initial Weight  Lowest Weight  \
Access Code                                                             
Y9RY2VSI     김승혜  FEMALE  25.0   172.0            66.9           65.8   
3GTN3S3B     허승준    MALE  26.0   176.0            70.0            NaN   
6B0IG276     이지민  FEMALE  23.0   171.0            98.0            NaN   
EMGRU2MO     장설윤  FEMALE  20.0   160.0            70.7            NaN   
1ELG96TX     서성빈  FEMALE  28.0   165.0            55.5            NaN   

             Target Weight     Status   Price            Purchased At  \
Access Code                                                             
Y9RY2VSI         55.000000  completed  112500 2017-04-14 19:03:29.976   
3GTN3S3B         65.000000  completed   44780 2017-05-23 20:53:54.368   
6B0IG276         91.140000  completed  132000 2017-08-23 23:39:21.840   
EMGRU2MO         53.000000  completed  112500 2017-08-28 20:18:22.824   
1ELG96TX         51.615002  completed   44780 2017-05-07 17:50:30.944   

              Channel  
Access Code            
Y9RY2VSI       others  
3GTN3S3B       others  
6B0IG276     facebook  
EMGRU2MO        naver  
1ELG96TX     facebook

In [126]:
raw_data["Gender"].unique()

array(['FEMALE', 'MALE', nan], dtype=object)

In [127]:
# data["Gender(clean)"] = data['Gender'].replace("MALE", "male").replace("FEMALE", "female")
data['Gender(clean)'] = data['Gender'].str.lower()
print(data.shape)
data[['Gender', 'Gender(clean)']].head()

(10000, 12)


Gender Gender(clean)
Access Code                      
Y9RY2VSI     FEMALE        female
3GTN3S3B       MALE          male
6B0IG276     FEMALE        female
EMGRU2MO     FEMALE        female
1ELG96TX     FEMALE        female

In [128]:
data['Gender(clean)'].value_counts()

female    8846
male      1023
Name: Gender(clean), dtype: int64

### 결측데이터 수정하기

In [129]:
raw_data["Height"].min()

-1.0

In [130]:
import numpy as np
data["Height(clean)"] = data["Height"].replace(-1, np.nan)

print(data.shape)
data.loc[data["Height"] <0, ["Name","Height","Height(clean)" ]]

(10000, 13)


Name  Height  Height(clean)
Access Code                            
O4OWMJG7     오세윤    -1.0            NaN
H6EV5AXL     박슬지    -1.0            NaN
O1IAZS7A     고솔윤    -1.0            NaN
5NEQOWHW     손초영    -1.0            NaN
OFAXUNXD     백채우    -1.0            NaN
8WGI0BO4     손슬윤    -1.0            NaN
C3RFRXK4     류서희    -1.0            NaN
TVISCJPR     유수민    -1.0            NaN
5DUJP5CO     배슬은    -1.0            NaN
P2YEDPO0     송설빈    -1.0            NaN
J9NO1247     문소빈    -1.0            NaN
XQ910MF8     오서주    -1.0            NaN
THMCHWBO     박설민    -1.0            NaN
0P0AORKD     서승이    -1.0            NaN
MEPHVEG6     한슬윤    -1.0            NaN
SDF5AWP3     유채유    -1.0            NaN
GB23YUBT     이승아    -1.0            NaN
888S8T3N     한성희    -1.0            NaN
GKRE4UB4     황서윤    -1.0            NaN
6R6CI3VP     권세은    -1.0            NaN
5URKM79N     허승연    -1.0            NaN

In [131]:
#다른방법
data["Height(clean)"] = data["Height"]
data.loc[data["Height"]<0, "Height(clean)"] = np.nan

In [132]:
data["Height(clean)"].min()

106.0

In [133]:
data["Height(clean)"].max()

203.2

In [134]:
data["Height(clean)"].mean()

163.54161860276196

In [135]:
pd.pivot_table(data, index = "Gender(clean)",values = "Height(clean)" )

Height(clean)
Gender(clean)               
female            162.116913
male              175.831965

In [136]:
raw_data["Age"].min()

0.0

In [137]:
raw_data["Age"].max()

173.0

In [138]:
data["Age(clean)"] = data["Age"]
data["Age(clean)"] = data["Age"].replace(0, np.nan)
data.loc[data["Age"]>=60, "Age(clean)"] = np.nan


In [139]:
data["Age(clean)"].mean()

27.39381024860477

In [140]:
data["Age(clean)"].max()

59.0

In [141]:
data["Age(clean)"].min()

13.0

In [142]:
pd.pivot_table(data, index = "Gender(clean)",values = "Age(clean)" )

Age(clean)
Gender(clean)            
female          27.172929
male            29.309127

### 특정 조건을 만족하는 고객 추출

In [143]:
new_columns = ['Name', 'Age(clean)', 'Height(clean)', 'Initial Weight',
    'Lowest Weight', 'Target Weight', 'Status']

new_data = data[new_columns].copy()
new_data

# print(data.shape)
# data.head()

Name  Age(clean)  Height(clean)  Initial Weight  Lowest Weight  \
Access Code                                                                  
Y9RY2VSI     김승혜        25.0          172.0            66.9           65.8   
3GTN3S3B     허승준        26.0          176.0            70.0            NaN   
6B0IG276     이지민        23.0          171.0            98.0            NaN   
EMGRU2MO     장설윤        20.0          160.0            70.7            NaN   
1ELG96TX     서성빈        28.0          165.0            55.5            NaN   
...          ...         ...            ...             ...            ...   
118AFCQ9     최서원        31.0          160.0            60.0            NaN   
VDE8FXV9     임세율        20.0          154.0            85.0           71.0   
VWJ4NLZY     정선희        28.0          166.0            69.0            NaN   
NJ2PR967     김유찬        28.0          188.0           110.0            NaN   
KVX64CPM     임수빈        26.0          167.0            57.5            NaN   

             Target Weight     Status  
Access Code                            
Y9RY2VSI         55.000000  completed  
3GTN3S3B         65.000000  completed  
6B0IG276         91.140000  completed  
EMGRU2MO         53.000000  completed  
1ELG96TX         51.615002  completed  
...                    ...        ...  
118AFCQ9         52.000000  completed  
VDE8FXV9         48.000000  completed  
VWJ4NLZY         54.000000  completed  
NJ2PR967         80.000000  cancelled  
KVX64CPM         50.500000  cancelled  

[10000 rows x 7 columns]

In [144]:
new_data['Weight_Loss(goal)']= new_data['Initial Weight'] - new_data['Target Weight']
new_data['Weight Loss(current)'] = new_data['Initial Weight'] - new_data['Lowest Weight']
new_data['BMI'] = (new_data['Initial Weight'])/((new_data['Height(clean)']/100)*(new_data['Height(clean)']/100))

new_data[['Name', 'Weight_Loss(goal)', 'Weight Loss(current)', 'BMI']].head()

Name  Weight_Loss(goal)  Weight Loss(current)        BMI
Access Code                                                         
Y9RY2VSI     김승혜          11.900000                   1.1  22.613575
3GTN3S3B     허승준           5.000000                   NaN  22.598140
6B0IG276     이지민           6.860000                   NaN  33.514586
EMGRU2MO     장설윤          17.700000                   NaN  27.617187
1ELG96TX     서성빈           3.884998                   NaN  20.385675

**잘못된 정보를 기입한 사용자(invalid user)를 찾기**

In [145]:
def invalid(x):
    Status = x["Status"]
    Age_clean = x["Age(clean)"]
    Height_clean = x['Height(clean)']
    Initial_Weight = x['Initial Weight']
    Lowest_Weight = x['Lowest Weight']
    Target_Weight = x['Target Weight']
    BMI = x['BMI']
    Weight_Loss_goal = x['Weight_Loss(goal)']
    
    if (Status == 'completed') and pd.isnull(Age_clean):
        return True
    elif (Status == 'completed') and pd.isnull(Height_clean):
        return True
    elif (Status == 'completed') and pd.isnull(Initial_Weight):
        return True
    elif (Status == 'completed') and pd.isnull(Lowest_Weight):
        return True
    elif (Status == 'completed') and pd.isnull(Target_Weight):
        return True
    elif (Status == 'completed') and (Height_clean <140 or Height_clean > 200) : 
        return True
    elif (Status == 'completed') and (BMI <18.5 or BMI > 30.0):
        return True
    elif (Status == 'completed') and Weight_Loss_goal <0:
        return True
    else:
        return False
    
new_data['invalid'] = new_data.apply(invalid, axis = 'columns')
new_data[(new_data)['invalid'] == True]



Name  Age(clean)  Height(clean)  Initial Weight  Lowest Weight  \
Access Code                                                                   
3GTN3S3B      허승준        26.0          176.0            70.0            NaN   
6B0IG276      이지민        23.0          171.0            98.0            NaN   
EMGRU2MO      장설윤        20.0          160.0            70.7            NaN   
1ELG96TX      서성빈        28.0          165.0            55.5            NaN   
U9AELHAT      배소정        22.0          161.0            76.0            NaN   
...           ...         ...            ...             ...            ...   
6FZ5OL4K     이 시민        19.0          176.0            78.1            NaN   
A8Z5J765      백성원        27.0          175.0            78.5            NaN   
118AFCQ9      최서원        31.0          160.0            60.0            NaN   
VDE8FXV9      임세율        20.0          154.0            85.0           71.0   
VWJ4NLZY      정선희        28.0          166.0            69.0            NaN   

             Target Weight     Status  Weight_Loss(goal)  \
Access Code                                                
3GTN3S3B         65.000000  completed           5.000000   
6B0IG276         91.140000  completed           6.860000   
EMGRU2MO         53.000000  completed          17.700000   
1ELG96TX         51.615002  completed           3.884998   
U9AELHAT         48.000000  completed          28.000000   
...                    ...        ...                ...   
6FZ5OL4K         72.633000  completed           5.467000   
A8Z5J765         70.500000  completed           8.000000   
118AFCQ9         52.000000  completed           8.000000   
VDE8FXV9         48.000000  completed          37.000000   
VWJ4NLZY         54.000000  completed          15.000000   

             Weight Loss(current)        BMI  invalid  
Access Code                                            
3GTN3S3B                      NaN  22.598140     True  
6B0IG276                      NaN  33.514586     True  
EMGRU2MO                      NaN  27.617187     True  
1ELG96TX                      NaN  20.385675     True  
U9AELHAT                      NaN  29.319856     True  
...                           ...        ...      ...  
6FZ5OL4K                      NaN  25.213068     True  
A8Z5J765                      NaN  25.632653     True  
118AFCQ9                      NaN  23.437500     True  
VDE8FXV9                     14.0  35.840783     True  
VWJ4NLZY                      NaN  25.039919     True  

[2619 rows x 11 columns]

In [146]:
def VIP(x):
    Status = x["Status"]
    BMI = x['BMI']
    Weight_Loss_goal = x['Weight_Loss(goal)']
    Weight_Loss_current = x['Weight Loss(current)']
    
    if (Status == 'completed') and ~pd.isnull(Weight_Loss_goal) and ~pd.isnull(Weight_Loss_current) and ~pd.isnull(BMI) :
        if  ~pd.isnull(Weight_Loss_current) :
            if Weight_Loss_current >= 10 :
                if BMI >= 30 :
                    if Weight_Loss_current > Weight_Loss_goal :
                        return True
    else:
        return False
    
new_data['VIP'] = new_data.apply(VIP, axis = 'columns')
new_data[(new_data)['VIP'] == True]




Name  Age(clean)  Height(clean)  Initial Weight  Lowest Weight  \
Access Code                                                                  
3T1I8I8E     임솔지        23.0          158.0         80.0137           54.5   
PJYKU9OW     홍윤오        31.0          174.0         99.9000           76.6   
0EMTSGLJ     류선정        34.0          167.0         86.0000           73.2   
FBEAIFW0     서서원        23.0          170.0         95.0000           75.7   
8QQV2YDW     홍서율        23.0          170.0         95.0000           75.7   
99KOLRU8     고서연        26.0          162.0        106.0000           95.2   
IBOWZ9WZ     손서애        22.0          166.0         83.0000           66.1   
6EH2LGR5     문세영        34.0          164.0         88.8000           56.3   
QQLYGTWD     황수윤        26.0          165.0        105.0000           87.8   
4Z1WB3UZ     허지예        26.0          162.0        106.0000           95.2   
2YAKET8R     서승희        30.0          164.0         99.8000           86.4   
LDPPDM0M     윤지안        19.0          168.0         93.1000           65.5   
PKHJWII8     정선영        23.0          170.0         95.0000           75.7   
3B3WQA4A     홍슬비        28.0          158.0         84.0000           69.3   
SDY4VS0P     오채현         NaN          170.0        103.1000           57.1   

             Target Weight     Status  Weight_Loss(goal)  \
Access Code                                                
3T1I8I8E         77.745742  completed           2.267958   
PJYKU9OW         84.000000  completed          15.900000   
0EMTSGLJ         80.050003  completed           5.949997   
FBEAIFW0         85.000000  completed          10.000000   
8QQV2YDW         85.000000  completed          10.000000   
99KOLRU8         98.580000  completed           7.420000   
IBOWZ9WZ         73.000000  completed          10.000000   
6EH2LGR5         82.584003  completed           6.215997   
QQLYGTWD         97.650000  completed           7.350000   
4Z1WB3UZ         98.580000  completed           7.420000   
2YAKET8R         92.814000  completed           6.986000   
LDPPDM0M         69.000000  completed          24.100000   
PKHJWII8         85.000000  completed          10.000000   
3B3WQA4A         75.000000  completed           9.000000   
SDY4VS0P        100.099998  completed           3.000001   

             Weight Loss(current)        BMI  invalid   VIP  
Access Code                                                  
3T1I8I8E                  25.5137  32.051634     True  True  
PJYKU9OW                  23.3000  32.996433     True  True  
0EMTSGLJ                  12.8000  30.836531     True  True  
FBEAIFW0                  19.3000  32.871972     True  True  
8QQV2YDW                  19.3000  32.871972     True  True  
99KOLRU8                  10.8000  40.390184     True  True  
IBOWZ9WZ                  16.9000  30.120482     True  True  
6EH2LGR5                  32.5000  33.016062     True  True  
QQLYGTWD                  17.2000  38.567493     True  True  
4Z1WB3UZ                  10.8000  40.390184     True  True  
2YAKET8R                  13.4000  37.105889     True  True  
LDPPDM0M                  27.6000  32.986111     True  True  
PKHJWII8                  19.3000  32.871972     True  True  
3B3WQA4A                  14.7000  33.648454     True  True  
SDY4VS0P                  46.0000  35.674740     True  True

VIP 사용자를 추출하여 추가 혜택, 홍보 등의 마케팅 방안을 제안할 수 있음.

## 결제 체크

  1. 한 명의 고객을 데려오는데 필요한 비용, 줄여서 고객 획득 비용(Customer Acquision Cost, 이하 **[CAC](https://neilpatel.com/blog/customer-acquisition-cost/)**)
  2. 한 명의 고객을 데려왔을 때, 고객이 회사에게 제공해주는 수익(Customer Lifetime Value, 이하 **[LTV](https://en.wikipedia.org/wiki/Customer_lifetime_value)**)
  
> LTV를 최대한 높이고, 동시에 CAC를 최대한 낮추는 쪽으로 서비스를 개선. 마케팅팀은 가능한 적은 비용을 지출하여 고객을 눔 코치에 유입시켜야 하며(CAC), 같은 CAC라면 이왕이면 회사에 많은 수익을 남겨주는 고객을 유입해야 합니다. (LTV)


  * LTV가 높은 고객군의 인구통계학적 정보를 이용해 페이스북 마케팅을 할 때 여성 고객들에게 집중적으로 광고를 보여주도록 타게팅 할 수 있음.
  * 요일/시간별 결제 비율. 가령 주중보다 주말에 결제할 확률이 높다면, 서비스를 유료로 결제할 의사가 있는 고객들에게 주말에 결제를 유도하는 메일을 보낼 수 있습니다.
  
이러한 요청을 종합하며, 마케팅팀의 의사결정에 도움이 될 수 있는 정보를 뽑아볼 수 있음.

**결제 / 캔슬 / 환불의 총 인원 수와 비율 구하기**

In [147]:
data.iloc[np.where(data['Age']<=17)[0]]

Name  Gender   Age  Height  Initial Weight  Lowest Weight  \
Access Code                                                             
GN4RTMY5     유서안  FEMALE  17.0   173.0            66.0           64.3   
4JR8819Q     배승희  FEMALE  17.0   162.0            57.0            NaN   
BV9EGOYI     황수연  FEMALE  16.0   163.0            59.0           58.8   
B5FA1TN7     임서혜  FEMALE  17.0   151.0            48.0           43.6   
WX3ORR7K     조소정  FEMALE  17.0   158.0            54.0           54.8   
...          ...     ...   ...     ...             ...            ...   
CDLV3AC4     권지연  FEMALE  17.0   162.0            61.0            NaN   
ASBOMYA9     한솔미  FEMALE  15.0   165.0            51.0            NaN   
00DT11XD     조세빈  FEMALE  16.0   159.0            51.0            NaN   
PCSD795E     홍슬윤  FEMALE  17.0   158.0            54.0           54.8   
ETT3AP6I     조승혁    MALE   0.0   170.0            70.0           68.2   

             Target Weight     Status   Price            Purchased At  \
Access Code                                                             
GN4RTMY5         48.000000  cancelled  112500 2017-05-08 00:39:26.688   
4JR8819Q         53.010002  completed  112500 2017-08-03 00:45:41.168   
BV9EGOYI         44.000000   refunded  112500 2017-08-05 12:04:25.928   
B5FA1TN7         44.000000   refunded  112500 2017-06-24 12:27:00.720   
WX3ORR7K         47.000000  completed  112500 2017-06-29 02:25:17.128   
...                    ...        ...     ...                     ...   
CDLV3AC4         54.000000  completed  112500 2017-07-25 09:06:31.656   
ASBOMYA9         45.000000  completed  112500 2017-08-14 18:20:49.336   
00DT11XD         48.000000  completed  112500 2017-07-22 01:51:55.200   
PCSD795E         47.000000  completed  112500 2017-06-29 02:22:19.128   
ETT3AP6I         65.000000  cancelled  112500 2017-08-02 23:11:49.912   

              Channel Gender(clean)  Height(clean)  Age(clean)  
Access Code                                                     
GN4RTMY5     facebook        female          173.0        17.0  
4JR8819Q       others        female          162.0        17.0  
BV9EGOYI        naver        female          163.0        16.0  
B5FA1TN7     facebook        female          151.0        17.0  
WX3ORR7K        naver        female          158.0        17.0  
...               ...           ...            ...         ...  
CDLV3AC4     facebook        female          162.0        17.0  
ASBOMYA9       others        female          165.0        15.0  
00DT11XD     facebook        female          159.0        16.0  
PCSD795E        naver        female          158.0        17.0  
ETT3AP6I     facebook          male          170.0         NaN  

[73 rows x 14 columns]

In [148]:
# .value_counts() 를 통해 Status의 종류와 그 누적 갯수를 출력합니다.
data["Status"].value_counts()

completed    5400
cancelled    4010
refunded      590
Name: Status, dtype: int64

**성별과 나이별 결제 / 캔슬 / 환불의 총 인원 수와 비율**

이렇게 인구통계학적 정보로 결제/캔슬/환불 비율을 확인하면, 마케팅팀에서는 가장 결제가 많이 일어나고 캔슬/환불이 적게 일어나는 성별과 나이에 마케팅 예산을 투입할 수 있음. 

In [149]:
def Age(x):
    age = x['Age(clean)']
    
    if 0 < age <= 17 :
        return "00 ~ 17"
    elif 18 <= age <= 24:
        return "18 ~ 24"
    elif 25 <= age <= 35:
        return "25 ~ 35"
    elif 36 <= age <= 44:
        return "36 ~ 44"
    elif 45 <= age <= 54:
        return "45 ~ 54"
    else:
        return "55 ~ 99"
    
   
data['Age(group)'] = data.apply(Age, axis = 'columns')
data.head()



Name  Gender   Age  Height  Initial Weight  Lowest Weight  \
Access Code                                                             
Y9RY2VSI     김승혜  FEMALE  25.0   172.0            66.9           65.8   
3GTN3S3B     허승준    MALE  26.0   176.0            70.0            NaN   
6B0IG276     이지민  FEMALE  23.0   171.0            98.0            NaN   
EMGRU2MO     장설윤  FEMALE  20.0   160.0            70.7            NaN   
1ELG96TX     서성빈  FEMALE  28.0   165.0            55.5            NaN   

             Target Weight     Status   Price            Purchased At  \
Access Code                                                             
Y9RY2VSI         55.000000  completed  112500 2017-04-14 19:03:29.976   
3GTN3S3B         65.000000  completed   44780 2017-05-23 20:53:54.368   
6B0IG276         91.140000  completed  132000 2017-08-23 23:39:21.840   
EMGRU2MO         53.000000  completed  112500 2017-08-28 20:18:22.824   
1ELG96TX         51.615002  completed   44780 2017-05-07 17:50:30.944   

              Channel Gender(clean)  Height(clean)  Age(clean) Age(group)  
Access Code                                                                
Y9RY2VSI       others        female          172.0        25.0    25 ~ 35  
3GTN3S3B       others          male          176.0        26.0    25 ~ 35  
6B0IG276     facebook        female          171.0        23.0    18 ~ 24  
EMGRU2MO        naver        female          160.0        20.0    18 ~ 24  
1ELG96TX     facebook        female          165.0        28.0    25 ~ 35

In [150]:
table = pd.pivot_table(data, index = ["Gender(clean)", 'Age(group)'], columns = 'Status', values = 'Age(clean)', aggfunc = 'count',
               fill_value = 0)

table["total"] = table['completed']+table['cancelled']+table['refunded']
table["conversion"] = table['completed']/table['total']

table

Status                    cancelled  completed  refunded  total  conversion
Gender(clean) Age(group)                                                   
female        00 ~ 17            25         35         3     63    0.555556
              18 ~ 24          1637       1827       149   3613    0.505674
              25 ~ 35          1664       2288       271   4223    0.541795
              36 ~ 44           206        421        46    673    0.625557
              45 ~ 54            74        160        25    259    0.617761
              55 ~ 99             0          5         0      5    1.000000
male          00 ~ 17             3          1         0      4    0.250000
              18 ~ 24            80        100        11    191    0.523560
              25 ~ 35           235        404        57    696    0.580460
              36 ~ 44            21         72         9    102    0.705882
              45 ~ 54             9         13         3     25    0.520000
              55 ~ 99             0          1         0      1    1.000000

분석 결과는 다음과 같습니다.

  * 결제가 가장 많이 일어나는 연령대 및 conversion 확인
  * 다만 이 채널들의 총 결제자(total)가 낮다는 것은 1) 아직 이 마케팅 채널이 최적화가 덜 되었거나, 2) 고객 획득 비용(CAC)이 높은 편이라 마케팅 비용을 늘리지 않았을 가능성이 있습니다. 또한 아주 희소한 경우이지만, 3) 주 마케팅 채널(ex: 페이스북)에 위 채널에 해당하는 고객의 인원수가 부족할 수도 있음.

이런 상황에서, 데이터분석가는 퍼포먼스 마케터와 함께 다음의 아이디어를 제시하여 회사의 매출을 증대할 수 있습니다.

  * 마케팅 예산을 여성 36 ~ 54세쪽에 집중한다. 이 채널이 전환율이 높기 때문에, CAC가 여성 25 ~ 35세와 동일하다면 여성 36 ~ 54세에 마케팅 예산을 늘리는 것은 좋은 전략입니다.
  * 여성 36 ~ 54세 채널의 CAC가 상대적으로 높다면, 이 CAC을 낮추는 시도를 합니다. 이 전략이 성공하면 그 후에 마케팅 예산을 집중하는 것도 방법입니다.
  * 현재 이용하고 있는 광고 채널을 다각화하여, 여성 36 ~ 54세가 활동하는 곳에 집중적으로 마케팅 예산을 투입하는 것도 시도해볼만 합니다.

**날짜와 요일 / 시간별 결제 / 캔슬 / 환불 비율**

이번에는 시간 정보를 기준으로 결제/캔슬/환불 비율을 알아보겠습니다. 크게 다음의 상황에 따른 구매 여부를 알고 싶습니다.

  1. 시간별 구매 현황(0시 ~ 23시)
  2. 요일별 구매 현황(월요일 ~ 일요일)
  
마케팅팀이 이 정보를 파악할 수 있다면, 1) 전환율이 높은 시기에 마케팅 예산 투입 비중을 줄이고/늘려서 CAC를 낮추거나, 2) 특정 시간대에 눔 코치의 유로 서비스를 아직 구매하지 않은 무료 사용자에게 유료 서비스 구매를 유도하는 메일을 보내서 매출을 늘릴 수 있음


분석 결과는 다음과 같습니다.

  * 아쉽게도, 구매 시간별 전환율(conversion)은 큰 차이가 없어 보입니다, 그 의미는 특정 시간대에 구매한 사용자들이 서비스를 이탈할 확률이 높아지거나 낮아지는 현상은 없다고 볼 수 있습니다.
  * 다만 전환율과는 별개로, 주로 점심시간(10시 ~ 12시)나 새벽(23시 ~ 24시)에 구매량이 대폭 늘어난다는 것을 알 수 있습니다. 만일 광고 예산을 집행한다면 이 시기에 집중적으로 집행하거나, 무료 사용자에게 유료 사용자로 전환을 유도하는 이메일을 보냄으로써 전환율을 높이는 것은 시도해볼만 합니다.

In [151]:
data["Purchased At"] = pd.to_datetime(data["Purchased At"])
data["Purchased At(hour)"] = data["Purchased At"].dt.hour


table_hour = pd.pivot_table(data, index = "Purchased At(hour)", columns = 'Status', values = 'Name', aggfunc = 'count',
               fill_value = 0)

table_hour['total'] = table_hour['completed']+table_hour['cancelled']+table_hour['refunded']
table_hour["conversion"] = table_hour['completed']/table_hour['total']

table_hour

Status              cancelled  completed  refunded  total  conversion
Purchased At(hour)                                                   
0                         236        344        26    606    0.567657
1                         156        207        28    391    0.529412
2                          90         97         4    191    0.507853
3                          58         66         5    129    0.511628
4                          59         45         7    111    0.405405
5                          36         47         6     89    0.528090
6                          48         70         6    124    0.564516
7                          80        114        20    214    0.532710
8                         171        264        29    464    0.568966
9                         162        239        36    437    0.546911
10                        208        323        38    569    0.567663
11                        212        263        27    502    0.523904
12                        205        235        36    476    0.493697
13                        205        286        41    532    0.537594
14                        192        253        20    465    0.544086
15                        187        231        14    432    0.534722
16                        187        235        26    448    0.524554
17                        180        246        26    452    0.544248
18                        194        260        20    474    0.548523
19                        163        269        36    468    0.574786
20                        184        236        22    442    0.533937
21                        231        329        32    592    0.555743
22                        248        332        41    621    0.534622
23                        318        409        44    771    0.530480

월요일-일요일 사이의 결제/캔슬/환불 비율은 다음의 방식으로 구할 수 있습니다.

In [152]:
import datetime

data["Purchased_At(weekday)_imsi"] = data["Purchased At"].dt.dayofweek
data['Purchased_At(weekday)_imsi'] = data['Purchased_At(weekday)_imsi'].astype('str')


def weekday(a):
    Purchased_At_weekday_imsi = a['Purchased_At(weekday)_imsi']
    
    if "0" in Purchased_At_weekday_imsi :
        return "Monday"
    elif "1" in Purchased_At_weekday_imsi:
        return "Tuesday"
    elif "2" in Purchased_At_weekday_imsi:
        return "Wednesday"
    elif "3" in Purchased_At_weekday_imsi:
        return "Thursday"
    elif "4" in Purchased_At_weekday_imsi:
        return "Friday"
    elif "5" in Purchased_At_weekday_imsi:
        return "Saturday"
    elif "6" in Purchased_At_weekday_imsi:
        return "Sunday"

data["Purchased At(weekday)"] = data.apply(weekday, axis = 'columns')


In [1]:
table_week = pd.pivot_table(data, index = "Purchased At(weekday)", columns = 'Status', values = 'Name', aggfunc = 'count',
               fill_value = 0)

table_week['total'] = table_week['completed']+table_week['cancelled']+table_week['refunded']
table_week["conversion"] = table_week['completed']/table_week['total']

table_week
# table_week = table_week[['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'], columns = 1]

NameError: name 'pd' is not defined

분석 결과는 다음과 같습니다.

  * 구매 시간과 마찬가지로, 구매 요일별 전환율(conversion)은 큰 차이가 없어 보입니다. 어느 요일이나 마찬가지로, 구매한 사람이 서비스를 이탈하거나 남을 확률은 거의 동일합니다.
  * 하지만 사용자들은 전반적으로 주말(금-일)이 다가올수록 구매를 덜 하게되고, 주중(월-수)이 다가올수록 구매를 많이 하게 되는 현상을 발견할 수 있습니다. 이 시기에 광고 예산을 크게 집행하거나, 구매를 유도하는 메일이나 모바일 노티피케이션을 보내는 것은 좋은 아이디어입니다.

**채널별 결제 / 캔슬 / 환불 비율**


In [154]:
data["Channel"].value_counts()

facebook     6880
others       1390
naver        1009
direct        297
email         271
google        120
instagram      33
Name: Channel, dtype: int64

이 채널별 마케팅 효율 정보를 알 수 있다면, 마케팅 팀에서 마케팅 예산을 재조정하여 1) 마케팅 효율이 좋은 채널에 예산을 집중하고, 2) 반대로 마케팅 효율이 좋지 않은 채널에 예산을 빼는 재조정(rebalancing)을 할 수 있습니다.

In [155]:
table_channel = pd.pivot_table(data, index = "Channel", columns = 'Status', values = 'Name', aggfunc = 'count',
               fill_value = 0)

table_channel['total'] = table_channel['completed']+table_channel['cancelled']+table_channel['refunded']
table_channel["conversion"] = table_channel['completed']/table_channel['total']


table_channel

Status     cancelled  completed  refunded  total  conversion
Channel                                                     
direct           119        169         9    297    0.569024
email             93        155        23    271    0.571956
facebook        2812       3654       414   6880    0.531105
google            42         66        12    120    0.550000
instagram         13         17         3     33    0.515152
naver            386        568        55   1009    0.562934
others           545        771        74   1390    0.554676

이 결과를 통해 알 수 있는 정보는 다음과 같습니다.

  * 현재 가장 많은 구매가 일어나는 채널은 페이스북(facebook) 입니다. 거의 대부분의 구매가 이 채널에서 일어났습니다.
  * 구매량이 100회 이상인 채널 중 가장 전환율이 높은 채널은 이메일(email) 입니다. 이 채널은 사용자가 웹사이트에 방문한 뒤, 바로 구매하지 않고 이메일 주소만만 남겨놨을 경우에 해당됩니다.
  * 아직 구매량이 페이스북만큼 많지는 않지만, 전환률이 페이스북보다 높은 채널 중 하나는 네이버(naver)입니다. 전환율이 56%로 페이스북보다 다소 높은 편입니다.
  * 네이버만큼이나 전환율이 높은 채널은 기타(others)입니다. 이 채널은 결제율이 페이스북만큼 높음에도 불구하고, 아쉽게도 기록이 잘 되어있지 않기 때문에 분석이 어렵습니다.
  
이 분석 결과를 통해 얻을 수 있는 아이디어는 다음과 같습니다.
  
  * 먼저 내부에서 트래킹 코드나 데이터 클리닝 코드를 수정하여, 기타(others) 채널을 더 세분화시킬 필요가 있습니다. 기타 채널은 1) 페이스북 만큼이나 구매량이 많으며, 2) 전환율이 페이스북보다 높습니다. 이 채널을 더 세분화시켜 분석한다면 마케팅 효율을 높일 수 있는 새로운 아이디어가 나올 수 있습니다.
  * 페이스북 다음으로 네이버 검색채널을 집중적으로 튜닝하거나 예산을 배정하여 마케팅 채널을 다각화할 수 있습니다.
  * 이메일(email)로 들어온 사용자가 전환율이 높은 이유를 더 분석할 수 있다면 좋겠습니다. 추측컨데, 눔 코치에 대한 신뢰도를 높일 다양한 정보를 이메일로 수신하였기 때문에 다른 채널에 비해 전환율이 높다는 가설을 세울 수 있습니다. 이 가설이 맞다면, 눔 코치를 이용하는 다른 사용자에게도 동일한 정보를 제공한다면 전체 전환율을 높일 수 있을 것입니다.